### DISCUSSION TITLE
How to keep suggestions inside the editor?

### DISCUSSION THREAD

**Discussion by krzychukula** [microsoft/monaco-editor#3631] [Q&A]
monaco-editor npm version: 0.7.0
Browser: Chrome
OS: OSX
How Can I keep suggestions inside editor? Suggestions should respect boundaries of the layout - not entire browser.

![screen shot 2016-12-21 at 11 18 32](https://cloud.githubusercontent.com/assets/240264/21386142/b79d4792-c771-11e6-8e80-5885c3c36867.png)

In this context I would like to see suggestions right corner to be aligned to the editor right corner.

Is it possible in current version as a flag or is there any way to manually fix suggestions window position?

**Comment by krzychukula** [comment]
I have tried:
`fixedOverflowWidgets: true` but it's not working or it's intended to do something different - not sure.


I have tried:
```
 for (const prop in this.editor.contentWidgets) {
      const contentWidget = this.editor.contentWidgets[prop];
      contentWidget.widget.allowEditorOverflow = false;
}
```
but it is used in `addWidget` so it's already to late:
```
if (widgetData.allowEditorOverflow) {
	this.overflowingContentWidgetsDomNode.appendChild(domNode);
} else {
	this.domNode.appendChild(domNode);
}
```

**Comment by krzychukula** [comment]
Additionally it seems inconsistent.
widgets default to true:
```
	// Editor.IContentWidget.allowEditorOverflow
	allowEditorOverflow = true;
```
while `addWidget` defaults to false:
```
allowEditorOverflow: widget.allowEditorOverflow || false,
```

**Comment by alexdima** [comment]
Each content widget can set `allowEditorOverflow` as it desires. The suggest widgets sets it to `true`.

We do not currently have an editor option to override/ignore that.

**Comment by krzychukula** [comment]
So let's treat it as feature request then.

**Comment by belohlavek** [comment]
Any updates?

**Comment by windischb** [comment]
Having the same problem...
Is there any workaround?

**Comment by krzychukula** [comment]
I used internals to move them to the dom root - but didn't update Monaco for some time already because of it

**Comment by stefan-lacatus** [comment]
> Each content widget can set `allowEditorOverflow` as it desires. The suggest widgets sets it to `true`.
> 
> We do not currently have an editor option to override/ignore that.

I can try to add a PR for this in vscode. Can you confirm that this is how you want it to work? 
 Have a flag on the editor called `keepWidgetsWithEditor` with 2 values:
    * `true` - all widgets will be forced to be drawn with the editor viewport. We do this by updating only `vs/editor/browser/viewParts/contentWidgets/contentWidgets.ts`
    * `false` - the default value, representing the current behavior

**Comment by krzychukula** [comment]
In my case problem was mostly because of z-index - suggestions were partly hidden under other UI elements.
1. Keeping them within UI of the editor would help but I feel it would impact UX (try it with an editor that has only 2 lines of text or even one)
2. What would work better in my case is what I did - move modals to the end of `<body>` to avoid all z-index issues by being last thing in DOM.

For (2) it could be optional field like:

```
idOfElementToPlaceModals: "domElementId"
```

**Comment by rbretoneiche** [comment]
Any Update ?

**Comment by cbn-falias** [comment]
For future readers:
If you're having troubles due to `z-index` or `overflow` like me, then [overflowWidgetsDomNode](https://microsoft.github.io/monaco-editor/typedoc/interfaces/editor.IEditorConstructionOptions.html#overflowWidgetsDomNode) might be a good alternative, as it allows you to render the widgets into your custom node (outside of the editor itself).

**Reply by JeanPerriault** [reply]
That's correct, but you'll have to copy a bunch of CSS + semantic CSS variables if you need to deal with dark theme, because for example "monaco-hover" and "monaco-editor-overlaymessage" are scoped with "monaco-editor" class, so they won't apply if you set overflowWidgetsDomNode to document.body for example. At least with version 0.45, don't know if this was/is the case with previous/new versions.
